engine/crates/engine-v2/walker/src/iter.rs

```diff
@@ -0,0 +1,144 @@
+use std::cmp::Ordering;
+
+use crate::{Walk, Walker};
+
+pub struct WalkIterator<'g, I, G> {
+    iter: I,
+    graph: &'g G,
+}
+
+impl<'g, I, G> Clone for WalkIterator<'g, I, G>
+where
+    I: Clone,
+{
+    fn clone(&self) -> Self {
+        Self {
+            iter: self.iter.clone(),
+            graph: self.graph,
+        }
+    }
+}
+
+impl<'g, I, G> Copy for WalkIterator<'g, I, G> where I: Copy {}
+
+impl<'g, I, G> WalkIterator<'g, I, G> {
+    pub fn new(iter: I, graph: &'g G) -> Self {
+        Self { iter, graph }
+    }
+}
+
+impl<'g, I, G> Iterator for WalkIterator<'g, I, G>
+where
+    I: Iterator,
+    <I as Iterator>::Item: Walk<G> + 'g,
+{
+    type Item = Walker<'g, <I as Iterator>::Item, G>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        self.iter.next().map(|item| item.walk(self.graph))
+    }
+}
+
+impl<'g, I, G> ExactSizeIterator for WalkIterator<'g, I, G>
+where
+    I: ExactSizeIterator,
+    <I as Iterator>::Item: Walk<G> + 'g,
+{
+    fn len(&self) -> usize {
+        self.iter.len()
+    }
+}
+
+impl<'g, I, G> std::iter::FusedIterator for WalkIterator<'g, I, G>
+where
+    I: std::iter::FusedIterator,
+    <I as Iterator>::Item: Walk<G> + 'g,
+{
+}
+
+impl<'g, I, G> std::iter::DoubleEndedIterator for WalkIterator<'g, I, G>
+where
+    I: std::iter::DoubleEndedIterator,
+    <I as Iterator>::Item: Walk<G> + 'g,
+{
+    fn next_back(&mut self) -> Option<Self::Item> {
+        self.iter.next_back().map(|item| item.walk(self.graph))
+    }
+}
+
+impl<'g, I, G> std::fmt::Debug for WalkIterator<'g, I, G>
+where
+    I: std::iter::ExactSizeIterator,
+    <I as Iterator>::Item: Walk<G> + 'g,
+    I: Clone,
+    <Self as Iterator>::Item: std::fmt::Debug,
+{
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_list().entries(self.clone()).finish()
+    }
+}
+
+impl<'g, I, G> std::cmp::Ord for WalkIterator<'g, I, G>
+where
+    I: std::iter::ExactSizeIterator,
+    <I as Iterator>::Item: Walk<G> + 'g,
+    I: Clone,
+    <Self as Iterator>::Item: std::cmp::Ord,
+{
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.len().cmp(&other.len()).then_with(|| {
+            for (left, right) in self.clone().zip(other.clone()) {
+                match left.cmp(&right) {
+                    Ordering::Equal => continue,
+                    other => return other,
+                }
+            }
+            Ordering::Equal
+        })
+    }
+}
```

Comment:
Not saying this is wrong, but an iterator implementing `Ord` seems a bit odd to me... Worth a comment explaining why it's needed if nothing else?

Reply:
Adding a comment later, I'm relying on it to compare field arguments in `@requires` since they can. This allows me to deduplicate/intern those required fields and make set manipulations at lot easier when determining the everything that is required during planning